### PR TITLE
DAOS-5657 vos: Add epoch uncertainty checks to VOS

### PR DESCRIPTION
**Pull request by jolivier23**
Adds epoch uncertainty checks to VOS
- adds write timestamps to the timestamp cache
- Add uncertainty tests to vts_mvcc.c

Two write timestamps are added.  Otherwise, we would
have to reject every operation that has a later write
regardless of whether it's inside the uncertainty
bound.

Note, this can potentially be improved a bit for
query and list cases as the checks at present
are done at a higher level than necessary.  For
example, for object listing, the check is done
on the container rather than each object.  This
should cover the objects but is a bit of a
heavy hammer.   But it's a line in the sand.

Co-authored-by: Li Wei <wei.g.li@intel.com>
Signed-off-by: Jeff Olivier <jeffrey.v.olivier@intel.com>

### CODE REVIEW

**Review by ryon-jensen — APPROVED** [review]
LGTM

File: src/vos/vos_ts.h [hunk 1]
```diff
@@ -760,6 +813,46 @@ vos_ts_set_update(struct vos_ts_set *ts_set, daos_epoch_t read_time)
...
+ *  \param[in]	write_time	The new write timestamp
+ */
+static inline void
+vos_ts_set_wupdate(struct vos_ts_set *ts_set, daos_epoch_t write_time)
+{
+	struct vos_ts_set_entry	*se;
+	int			 i;
+
+	if (!vos_ts_in_tx(ts_set))
+		return;
```
Comment (liw) [inline src/vos/vos_ts.h:845]: An epoch uncertainty check also looks for standalone writes. Yeah, we should add those cases to vts_mvcc.

File: src/vos/vos_ts.h [hunk 2]
```diff
@@ -62,6 +62,13 @@ struct vos_ts_pair {
 	struct dtx_id	tp_tx_rh;
 };
 
+struct vos_wts_cache {
+	/** write timestamps */
+	daos_epoch_t	wc_ts_w[2];
```
Comment (liw) [inline src/vos/vos_ts.h:67]: Style: Some comments defining what do "min" and "max" do would be helpful.

File: src/vos/vos_ts.h [hunk 3]
```diff
@@ -62,6 +62,13 @@ struct vos_ts_pair {
...
+	/** Index of max timestamp */
+	uint32_t	wc_w_max;
```
Comment (liw) [inline src/vos/vos_ts.h:69]: Style: "/* Index of max timestamp in wc_ts_w */".

File: src/vos/vos_ts.h (hunk 2 above)
Comment (liw) [reply src/vos/vos_ts.h:67]: "min" actually means "second highest", not "minimum"?

File: src/vos/vos_ts.h [hunk 4]
```diff
@@ -421,6 +432,48 @@ vos_ts_get_negative(struct vos_ts_set *ts_set, uint64_t hash, bool reset)
...
+	max = wcache->wc_ts_w[max_idx];
+	if (epoch >= max)
+		return false;
+
+	min = wcache->wc_ts_w[1 - max_idx];
+	if (epoch < min) /* unknown case, not enough history */
+		return true;
+
+	if (bound >= max) /* write conflict */
+		return true;
```
Comment (liw) [inline src/vos/vos_ts.h:473]: Did you mean

    if (bound >= min)

instead?

File: src/vos/vos_ts.h (hunk 1 above)
Comment (jolivier23) [reply src/vos/vos_ts.h:845]: There are no such cases in practice.   Even solo writes pass in a DTX so from VOS perspective, they look like a transaction.

File: src/vos/vos_ts.h (hunk 4 above)
Comment (jolivier23) [reply src/vos/vos_ts.h]: no, I added a comment inline in the code...is it clearer ?

File: src/vos/vos_ts.h (hunk 3 above)
Comment (jolivier23) [reply src/vos/vos_ts.h]: done

File: src/vos/vos_ts.h (hunk 2 above)
Comment (jolivier23) [reply src/vos/vos_ts.h]: correct, I changed the names.   Hopefully it's more clear

File: src/vos/vos_ts.h (hunk 3 above)
Comment (ryon-jensen) [reply src/vos/vos_ts.h]: Curious what the style guideline for comments actually is. I see both '/**' and '/*' in the source.

**Review by ryon-jensen — APPROVED** [review]

**Review by ryon-jensen — APPROVED** [review]

File: src/vos/vos_ts.h (hunk 1 above)
Comment (liw) [reply src/vos/vos_ts.h:871]: Yes, indeed. Sorry.

File: src/vos/vos_ts.h (hunk 2 above)
Comment (liw) [reply src/vos/vos_ts.h]: They are much easier to understand now. Thanks.

File: src/vos/vos_ts.h (hunk 4 above)
Comment (liw) [reply src/vos/vos_ts.h]: Yes, much clearer. Thanks.

File: src/vos/vos_ts.h (hunk 3 above)
Comment (liw) [reply src/vos/vos_ts.h]: I don't know the "official" one, but my understanding is that documentation for APIs shall be Doxygen, whereas non-API comments (e.g., one in the middle of a function and one for a private field) shall not be Doxygen.

Comment (liw) [reply src/vos/vos_ts.h]: Oh, my style comment above isn't about Doxygen, but "in wc_ts_w", as I initially didn't know it's an index of what data structure.

**Review by liw — APPROVED** [review]
Looks good to me from a high level; not really familiar with vos code.

**Review by NiuYawei — APPROVED** [review]

**Pull request merged**